ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouterMetrics.java

```diff
@@ -99,6 +204,249 @@ public Integer getValue() {
         return currentOperationsCount.get();
       }
     };
-    metricRegistry.register(MetricRegistry.name(NonBlockingRouter.class, "-NumActiveOperations"), activeOperations);
+    metricRegistry.register(MetricRegistry.name(NonBlockingRouter.class, "NumActiveOperations"), activeOperations);
+  }
+
+  /**
+   * Count error based on error type.
+   * <p/>
+   * This method should be called when an {@code Operation} is completed or aborted.
+   * @param exception The exception to be counted.
+   */
+  void countError(Exception exception) {
+    operationErrorRate.mark();
+    if (exception instanceof RouterException) {
+      switch (((RouterException) exception).getErrorCode()) {
+        case AmbryUnavailable:
+          ambryUnavailableErrorCount.inc();
+          break;
+        case InvalidBlobId:
+          invalidBlobIdErrorCount.inc();
+          break;
+        case InvalidPutArgument:
+          invalidPutArgumentErrorCount.inc();
+          break;
+        case OperationTimedOut:
+          operationTimedOutErrorCount.inc();
+          break;
+        case RouterClosed:
+          routerClosedErrorCount.inc();
+          break;
+        case UnexpectedInternalError:
+          unexpectedInternalErrorCount.inc();
+          break;
+        case BlobTooLarge:
+          blobTooLargeErrorCount.inc();
+          break;
+        case BadInputChannel:
+          blobTooLargeErrorCount.inc();
+          break;
+        case InsufficientCapacity:
+          insufficientCapacityErrorCount.inc();
+          break;
+        case BlobDeleted:
+          blobDeletedErrorCount.inc();
+          break;
+        case BlobDoesNotExist:
+          blobDoesNotExistErrorCount.inc();
+          break;
+        case BlobExpired:
+          blobExpiredErrorCount.inc();
+          break;
+        default:
+          unknownErrorCountForOperation.inc();
+          break;
+      }
+    } else {
+      unknownErrorCountForOperation.inc();
+    }
+  }
+
+  /**
+   * Get {@link NodeLevelMetrics} for a given {@link DataNodeId}. If the {@link DataNodeId} as the key does not exist,
+   * a {@link NodeLevelMetrics} corresponding to an unknown {@link DataNodeId} will be returned.
+   * @param dataNodeId The {@link DataNodeId} to be indexed.
+   * @return The {@link NodeLevelMetrics}.
+   */
+  NodeLevelMetrics getDataNodeBasedMetrics(DataNodeId dataNodeId) {
+    return dataNodeToMetrics.containsKey(dataNodeId) ? dataNodeToMetrics.get(dataNodeId) : dataNodeToMetrics.get(null);
```

Comment:
Theoretically, an `unknown` datanode should never be returned, but . If for any reason a non-existent `dataNodeId` is passed into this method, a `null` will be returned and the caller should handle this abnormal case, or record this in the metrics. I just feel the caller-handling logic will be needed at many places and introduce more code, while by updating metrics for an `unknown` datanode still allows us to understand something unexpected happened.